### ISSUE TITLE
Send the article to early readers

### ISSUE THREAD

**Issue by Witiko** [Witiko/fast-regression-testing#10]
After closing #7, #8, and #9, we should send out the draft of the article to early readers. Here is a list of early readers who might be interested in reading the article and providing feedback:

- Petr Sojka, a PhD supervisor and mentor of @Witiko.
- Joseph Wright, the main contributor of the l3build package
- Andrej Genčur, the second main contributor to the Markdown package since version 3.0.0.
- Tereza Vrabcová, Dominik Rehák, and Michal Hoftich, contributors to the Markdown package.

Furthermore, Marei mentioned that we could also reach out to members of the IslandOfTeX's TeX Infrastructure working group:

- Oliver Kopp
- Patrick Gundlach
- Ben Frank

Lastly, Marei also mentioned a number of Sysamin/DevOps friends, who are less knowledgeable about TeX, but who may provide useful feedback regarding existing testing frameworks and the techniques used in these frameworks.

**Comment by Witiko** [comment]
From witiko@mail.muni.cz Wed Dec  6 11:23:38 2023  
Date: Wed, 6 Dec 2023 11:23:38 +0100  
To: Joseph Wright <joseph.wright@morningstar2.co.uk>  
Subject: Re: Invitation for feedback on TUGboat article: "Fast Regression Testing of TeX Packages"  

Dear Joseph,

thank you for your valuable feedback. I respond to your comments below.

> - What defines a 'test' is rather tricky - I guess like me you are likely
> counting test *files* - perhaps worth saying? (One file can contain many
> tests - see the latex3 set or those I've got for siunitx, for example)

In the Markdown package, a test file is synonymous with a test.
Every test file contains only one test case, as shown in Section 2.1.

> - I wonder why your tests take so long: I see you have around 783 test
> (files?) but these are taking 15 h before optimisation. In contrast, for the
> LaTeX kernel we have just short of 600 tests for base, and they run with
> three engines, taking 40-ish minutes on GitHub Actions. The whole LaTeX
> kernel build process takes just over an hour when it runs for the CTAN
> target, which is a linear approach (no breaking up into units)

We have 783 test files and perform 14 runs of TeX with different TeX
formats, engines, and templates (see Section 2.2). This means that we
need 10,962 runs of TeX to process all test files in GitHub Actions,
see also Section 5. If the tests take 15 hours, then an average run of
TeX takes 4.93 seconds, which seems reasonable.

> - You've not said why you picked Python

Using Python instead of Bash or Lua made it possible to implement
the batching of test files and multiprocessing. Implementing both
batching and multiprocessing within our framework without reliance on
external tools allowed us to also implement techniques such as batch
size limiting (see Section 6.3). Below, I explain why either Bash or Lua
would be poor choices.

Whereas Bash is convenient for simple scripts, more complicated
algorithms seem best written in a more expressive language such as
Lua or Python. Whereas Lua can be as expressive as Python, it has no
built-in support for multiprocessing, as we also mentioned in Section
7.1.  Furthermore, Python's type system allows us to type-check our code
in GitHub Actions, which we could not do with Lua. This gives us useful
safety net when we make changes to our framework.

> Lua is used by l3build as its a certainty for any TeX system, whereas
> Python ... isn't

I realize that we mentioned platform-independence as the
reason/consequence of using Python in the last paragraph of Section 4.1.
Although Python is platform-independent, it was not our reason for
picking Python and we will rewrite the paragraph to reflect that.

We don't feel as strongly about the portability of our test framework,
because it is not part of the runtime of the Markdown package.
Therefore, we feel it's reasonable to assume that the tests run in an
environment where both TeX and Python are available, especially given
how ubiquitous Python has become.

> - You mention cross-platform working, but in the script you show (before
> changes), you are using diff with a raw log - could you say how you handle
> that for your new setup to normalise out platform/engine/format differences?

As you can see in Section 2.1, the expected output does not cover the
entire .log file, just the parts produced by the markdown commands.
For example, the markdown text `Hello ~~world~~!` will produce the
following output in the .log file:

    TEST INPUT BEGIN
    strikeThrough: world
    TEST INPUT END

Then, the framework only considers the inside of these `TEST INPUT`
sections, which helps smooth over the platform/engine/format
differences.

> - Do you get any issues of interference between batches? As TeX writes a
> load of files, we have to be pretty careful to avoid one run influencing
> another, which sounds harder if they are happening at the same time.

Since our tests only validate the conversion from Markdown to TeX, they
do not engage many other mechanisms present in TeX and the different TeX
formats. Therefore, TeX groups are sufficient to isolate the individual
test files in a batch.

Best,

Vitek

**Comment by Witiko** [comment]
On December 11, we received feedback from Ben Frank. It can be found in a commented PDF document `markdown-fast-regression-testing_feedback_ben.pdf` in my exchange with @TeXhackse on Matrix.org from the same day.

On December 12, I responded to the feedback from Ben Frank. My response can be found in a commented PDF document `markdown-fast-regression-testing_feedback_ben_response.pdf` in my exchange with @TeXhackse on Matrix.org from the same day.

On December 14, we exchanged further comments with Ben Frank on Matrix.org.
